blockdag/process.go

```diff
@@ -153,60 +30,82 @@ func (dag *BlockDAG) ProcessBlock(block *util.Block, flags BehaviorFlags) (isOrp
 }
 
 func (dag *BlockDAG) processBlockNoLock(block *util.Block, flags BehaviorFlags) (isOrphan bool, isDelayed bool, err error) {
-	isAfterDelay := flags&BFAfterDelay == BFAfterDelay
-	wasBlockStored := flags&BFWasStored == BFWasStored
-	disallowDelay := flags&BFDisallowDelay == BFDisallowDelay
-	disallowOrphans := flags&BFDisallowOrphans == BFDisallowOrphans
-
 	blockHash := block.Hash()
 	log.Tracef("Processing block %s", blockHash)
 
-	// The block must not already exist in the DAG.
-	if dag.IsInDAG(blockHash) && !wasBlockStored {
-		str := fmt.Sprintf("already have block %s", blockHash)
-		return false, false, ruleError(ErrDuplicateBlock, str)
+	err = dag.checkDuplicateBlock(blockHash, flags)
+	if err != nil {
+		return false, false, err
+	}
+
+	err = dag.checkBlockSanity(block, flags)
+	if err != nil {
+		return false, false, err
+	}
+
+	isOrphan, isDelayed, err = dag.processBlockCheckDelayedAndOrphanBlocks(block, flags)
+	if isOrphan || isDelayed || err != nil {
+		return isOrphan, isDelayed, err
 	}
 
-	// The block must not already exist as an orphan.
-	if _, exists := dag.orphans[*blockHash]; exists {
-		str := fmt.Sprintf("already have block (orphan) %s", blockHash)
-		return false, false, ruleError(ErrDuplicateBlock, str)
+	err = dag.maybeAcceptBlock(block, flags)
+	if err != nil {
+		return false, false, err
 	}
 
-	if dag.isKnownDelayedBlock(blockHash) {
-		str := fmt.Sprintf("already have block (delayed) %s", blockHash)
-		return false, false, ruleError(ErrDuplicateBlock, str)
+	err = dag.processOrphansAndDelayedBlocks(blockHash, flags)
+	if err != nil {
+		return false, false, err
 	}
 
-	if !isAfterDelay {
-		// Perform preliminary sanity checks on the block and its transactions.
-		delay, err := dag.checkBlockSanity(block, flags)
+	log.Debugf("Accepted block %s", blockHash)
+
+	return false, false, nil
+}
+
+func (dag *BlockDAG) processBlockCheckDelayedAndOrphanBlocks(block *util.Block, flags BehaviorFlags) (isOrphan bool, isDelayed bool, err error) {
```

Comment:
Why the `processBlock` prefix here and in `processBlockCheckBlockDelay` and `processBlockCheckMissingParents`?

Reply:
processBlockCheckMissingParents doesn't have a corresponding something in validate.go, so I think it can be safely renamed to checkMissingParents.

So we are left with processBlockCheckBlockDelay and processBlockCheckDelayedAndOrphanBlocks the can be confused with checkBlockDelayed

I suggest to rename checkBlockDelayed to shouldBlockBeDelayed.
Then remove prefix from processBlockCheckBlockDelay and processBlockCheckDelayedAndOrphanBlocks